pkg/volume/host_path/host_path.go

```diff
@@ -314,3 +318,184 @@ func getVolumeSource(spec *volume.Spec) (*v1.HostPathVolumeSource, bool, error)
 
 	return nil, false, fmt.Errorf("Spec does not reference an HostPath volume type")
 }
+
+type hostPathTypeChecker interface {
+	Exists() bool
+	IsFile() bool
+	MakeFile() error
+	IsDir() bool
+	MakeDir() error
+	IsBlock() bool
+	IsChar() bool
+	IsSocket() bool
+	GetPath() string
+}
+
+type fileTypeChecker interface {
+	getFileType(fileInfo os.FileInfo) (v1.HostPathType, error)
+}
+
+// this is implemented in per-OS files
+type defaultFileTypeChecker struct{}
+
+type osFileTypeChecker struct {
+	path    string
+	exists  bool
+	info    os.FileInfo
+	checker fileTypeChecker
+}
+
+func (ftc *osFileTypeChecker) Exists() bool {
+	return ftc.exists
+}
+
+func (ftc *osFileTypeChecker) IsFile() bool {
+	if !ftc.Exists() {
+		return false
+	}
+	return !ftc.info.IsDir()
+}
+
+func (ftc *osFileTypeChecker) MakeFile() error {
+	f, err := os.OpenFile(ftc.path, os.O_CREATE, os.FileMode(0644))
+	defer f.Close()
+	if err != nil {
+		if !os.IsExist(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ftc *osFileTypeChecker) IsDir() bool {
+	if !ftc.Exists() {
+		return false
+	}
+	return ftc.info.IsDir()
+}
+
+func (ftc *osFileTypeChecker) MakeDir() error {
+	err := os.MkdirAll(ftc.path, os.FileMode(0755))
+	if err != nil {
+		if !os.IsExist(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ftc *osFileTypeChecker) IsBlock() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	blkDevType, err := ftc.checker.getFileType(ftc.info)
+	if err != nil {
+		return false
+	}
+	return blkDevType == v1.HostPathBlockDev
+}
+
+func (ftc *osFileTypeChecker) IsChar() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	charDevType, err := ftc.checker.getFileType(ftc.info)
+	if err != nil {
+		return false
+	}
+	return charDevType == v1.HostPathCharDev
+}
+
+func (ftc *osFileTypeChecker) IsSocket() bool {
+	if !ftc.Exists() {
+		return false
+	}
+
+	socketType, err := ftc.checker.getFileType(ftc.info)
+	if err != nil {
+		return false
+	}
+	return socketType == v1.HostPathSocket
+}
+
+func (ftc *osFileTypeChecker) GetPath() string {
+	return ftc.path
+}
+
+func newOSFileTypeChecker(path string, checker fileTypeChecker) (hostPathTypeChecker, error) {
+	ftc := osFileTypeChecker{path: path, checker: checker}
+	info, err := os.Stat(path)
+	if err != nil {
+		ftc.exists = os.IsExist(err)
```

Comment:
@euank Right.

Also since the `err` of `os.Stat` can be non nil, so there is no need to check what the error is. Because `info` is nil.

I think just return nil for non-nil error of `os.Stat`.